### PR TITLE
Add token to login

### PR DESCRIPTION
**Pull request by simionescuraluca**

### CODE REVIEW

File: src/main/java/com/vinyl/model/Token.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	private Integer id;
+	
+	@Column(name="HASH")
+	@NotNull
+	private String hash;
+
+	@OneToOne
+	@NotNull
+	@JoinColumn(name="user_id")
+	private User user;
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/model/Token.java:37]: add a validUntil field, token should not be valid forever, make it valid maybe for 1 month after it is created

File: src/main/java/com/vinyl/model/Token.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToOne;
+import javax.validation.constraints.NotNull;
+
+import org.hibernate.annotations.GenericGenerator;
+
+
+@Entity
+public class Token {
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/model/Token.java:19]: a liquibase script that creates the table?

File: src/main/java/com/vinyl/model/Token.java (hunk 1 above)
Comment (dragosslujeru) [inline src/main/java/com/vinyl/model/Token.java:37]: I think the relationship here could be ManyToOne (one user will have many tokens as after one expires he will create a new one)

File: src/main/java/com/vinyl/service/UserService.java [hunk 3]
```diff
@@ -59,10 +64,23 @@ public void deleteUser(EmailPassDTO credentials) {
...
+		
+		token.setUser(userRepository.findByEmail(info.getEmail()).get());
+		token.setHash(String.valueOf(Math.abs(info.hashCode())));
+
+		return save(token);
+	}
 
+	public Token save(Token toSave) {
+		Token token = tokenRepository.findByHash(toSave.getHash());
+		if (token == null) {
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/service/UserService.java:75]: if token is null you try to save the null in the db?

File: src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java [hunk 4]
```diff
@@ -13,6 +13,8 @@
 
 	private final UserRepository userRepository;
 	private final BCryptPasswordEncoder passwordEncoder;
+	
+	private User user;
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java]: you dont seem to be using this field, remove it

File: src/main/java/com/vinyl/modelDTO/TokenDTO.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.vinyl.modelDTO;
+
+public class TokenDTO {
+	
+	private String hash;
+
+
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/modelDTO/TokenDTO.java]: remove extra line

File: src/main/java/com/vinyl/model/Token.java (hunk 1 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/model/Token.java:37]: done

File: src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java (hunk 4 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java]: done

File: src/main/java/com/vinyl/modelDTO/TokenDTO.java (hunk 5 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/modelDTO/TokenDTO.java]: done

File: src/main/java/com/vinyl/model/Token.java (hunk 1 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/model/Token.java:37]: done

File: src/main/java/com/vinyl/model/Token.java (hunk 2 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/model/Token.java:19]: done

File: src/main/java/com/vinyl/service/UserService.java (hunk 3 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/service/UserService.java:75]: if token is not in the database, then i save the toSave token given by the param.
so if token is already in the database ( token != null ) i just return it. 
the token variable is actually the stored token in the database.

Comment (dragosslujeru) [reply src/main/java/com/vinyl/service/UserService.java:75]: ok, then it is a little wierd, you do a lot of work (creating the token, populating it's fields, etc) and then you just discard that work if a token already exists in the db.
You should check if a token exists at the beginning of the login method, then check if it is still valid (current time is not greater than validUntil) if it is return in, otherwise create a new one and return it

File: src/main/java/com/vinyl/model/Token.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
+	@GenericGenerator(name = "native", strategy = "native")
+	private Integer id;
+	
+	@Column(name="HASH")
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/model/Token.java]: be consistent with column naming (either all caps or all lowercase

File: src/main/java/com/vinyl/repository/TokenRepository.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.stereotype.Repository;
+
+import com.vinyl.model.Token;
+import com.vinyl.model.User;
+
+@Repository
+public interface TokenRepository extends CrudRepository<Token, Integer>{
+
+	Token findFirstByUserOrderByValidUntilDesc(User user);
+	Token findByHash(String hash);
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/repository/TokenRepository.java]: this method is not used

File: src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java [hunk 8]
```diff
@@ -13,6 +13,8 @@
 
 	private final UserRepository userRepository;
 	private final BCryptPasswordEncoder passwordEncoder;
+
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java]: unrelated changes

File: src/test/java/com/vinyl/vinylstore/LoginTest.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	@Autowired
+	private TokenRepository tokenRepository;
+
+	@Autowired
+	private BCryptPasswordEncoder passwordEncoder;
+
+	@Autowired
+	private ValidatorFactory validatorFactory;
+
+	private BCryptPasswordEncoder mockedEncoder;
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/LoginTest.java]: rename to mockedPasswordEncoder

File: src/main/java/com/vinyl/model/Token.java (hunk 6 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/model/Token.java]: fixed

File: src/main/java/com/vinyl/repository/TokenRepository.java (hunk 7 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/repository/TokenRepository.java]: fixed

File: src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java (hunk 8 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/service/validation/EmailAndPasswordValidator.java]: fixed

File: src/test/java/com/vinyl/vinylstore/LoginTest.java (hunk 9 above)
Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/LoginTest.java]: done

File: src/test/java/com/vinyl/vinylstore/LoginTest.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	public void tearDown() {
+		tokenRepository.deleteAll();
+		userRepository.deleteAll();
+		addressRepository.deleteAll();
+		roleRepository.deleteAll();
+		validatorFactory.setPasswordEncoder(passwordEncoder);
+	}
+
+	@Test
+	public void loginWithValidEmailAndPassword() {
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/LoginTest.java]: this test is for loginWithValidEmailAndPass and when token already exists, rename it as such and add a new one when token does not exists and it will be created when the request is done

File: src/test/java/com/vinyl/vinylstore/Test.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    @Autowired
+    RoleRepository roleRepository;
+
+    @Autowired
+    AddressRepository addressRepository;
+
+    @Autowired
+    UserRepository userRepository;
+
+    abstract Address createAddress();
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/Test.java]: you don't need to have themse method abstract, move the code from the subclass here

File: src/test/java/com/vinyl/vinylstore/Test.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureTestDatabase
+@RunWith(SpringRunner.class)
+@ActiveProfiles("test")
+public abstract class Test {
+    @Autowired
+    TestRestTemplate trt;
+
+    @Autowired
+    RoleRepository roleRepository;
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/Test.java]: it would make more sense if these fields would be made protected (they can used by subclasses)

File: src/test/java/com/vinyl/vinylstore/Test.java (hunk 11 above)
Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/Test.java]: fixed

File: src/test/java/com/vinyl/vinylstore/Test.java (hunk 12 above)
Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/Test.java]: fixed

File: src/test/java/com/vinyl/vinylstore/LoginTest.java (hunk 10 above)
Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/LoginTest.java]: done

File: src/test/java/com/vinyl/vinylstore/LoginTest.java [hunk 13]
```diff
@@ -57,19 +56,26 @@ public void tearDown() {
...
 	}
 
+	@Test
+	public void loginWithValidEmailAndPasswordWhenTokenDoesNotExist() {
+		setRequest(request);
+
+		ResponseEntity<TokenDTO> tdo = trt.postForEntity("/users/login", request, TokenDTO.class);
+
+		Assertions.assertThat(tdo.getStatusCode()).isEqualTo(HttpStatus.OK);
+		Assertions.assertThat(tdo.getBody().getHash());
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/LoginTest.java]: you should actually do a find in the db and assert that the query is not null

Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/LoginTest.java]: done

File: src/main/java/com/vinyl/repository/TokenRepository.java [hunk 14]
```diff
@@ -10,4 +10,5 @@
 public interface TokenRepository extends CrudRepository<Token, Integer>{
 
 	Token findFirstByUserOrderByValidUntilDesc(User user);
+	Token findByUserId(Integer userId);
```
Comment (dragosslujeru) [inline src/main/java/com/vinyl/repository/TokenRepository.java:13]: the user and token has a 1 to many relationship, if you try to find tokens by a user id you will get by a list (not a single user). I think you can just use the existing method findFirstByUserOrderByValidUntilDesc to check if the token exists

File: src/test/java/com/vinyl/vinylstore/LoginTest.java [hunk 15]
```diff
@@ -73,7 +73,7 @@ public void loginWithValidEmailAndPasswordWhenTokenDoesNotExist() {
 		ResponseEntity<TokenDTO> tdo = trt.postForEntity("/users/login", request, TokenDTO.class);
 
 		Assertions.assertThat(tdo.getStatusCode()).isEqualTo(HttpStatus.OK);
-		Assertions.assertThat(tdo.getBody().getHash());
+		Assertions.assertThat(tokenRepository.findByUserId(user.getId()));
```
Comment (dragosslujeru) [inline src/test/java/com/vinyl/vinylstore/LoginTest.java:76]: Assertions.assertThat([The Object]).isNotNull();

**Review by dragosslujeru — APPROVED** [review]

File: src/main/java/com/vinyl/repository/TokenRepository.java (hunk 14 above)
Comment (simionescuraluca) [reply src/main/java/com/vinyl/repository/TokenRepository.java:13]: done

File: src/test/java/com/vinyl/vinylstore/LoginTest.java (hunk 15 above)
Comment (simionescuraluca) [reply src/test/java/com/vinyl/vinylstore/LoginTest.java:76]: done
